runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkMetricResults.java

```diff
@@ -0,0 +1,184 @@
+            @Override
+            public boolean apply(CounterAggregator metricResult) {
+              return matches(filter, metricResult.getKey());
+            }
+          });
+      return Iterables.transform(filteredCounters,
+          new Function<CounterAggregator, MetricResult<Long>>() {
+            @SuppressWarnings("ConstantConditions")
+            @Nullable
+            @Override
+            public MetricResult<Long> apply(CounterAggregator metricResult) {
+              MetricKey key = metricResult.getKey();
+              return new SparkMetricResult<>(key.metricName(), key.stepName(),
+                  metricResult.getValue());
+            }
+          });
+    }
+
+    @Override
+    public Iterable<MetricResult<DistributionResult>> distributions() {
+      Iterable<DistributionAggregator> allDistributions =
+          metricsAccum.value().getDistributions();
+      Iterable<DistributionAggregator> filteredDistributions =
+          Iterables.filter(allDistributions,
+              new Predicate<DistributionAggregator>() {
+            @SuppressWarnings("ConstantConditions")
+            @Override
+            public boolean apply(DistributionAggregator metricResult) {
+              return matches(filter, metricResult.getKey());
+            }
+          });
+      return Iterables.transform(filteredDistributions,
+          new Function<DistributionAggregator, MetricResult<DistributionResult>>() {
+            @SuppressWarnings("ConstantConditions")
+            @Nullable
+            @Override
+            public MetricResult<DistributionResult>
+            apply(DistributionAggregator metricResult) {
+              MetricKey key = metricResult.getKey();
+              return new SparkMetricResult<>(key.metricName(), key.stepName(),
+                  metricResult.getValue().extractResult());
+            }
+          });
+    }
+
+    private boolean matches(MetricsFilter filter, MetricKey key) {
+      return matchesName(key.metricName(), filter.names())
+          && matchesScope(key.stepName(), filter.steps());
+    }
+
+    private boolean matchesName(MetricName metricName, Set<MetricNameFilter> nameFilters) {
+      if (nameFilters.isEmpty()) {
+        return true;
+      }
+
+      for (MetricNameFilter nameFilter : nameFilters) {
+        if ((nameFilter.getName() == null || nameFilter.getName().equals(metricName.name()))
+            && Objects.equal(metricName.namespace(), nameFilter.getNamespace())) {
+          return true;
+        }
+      }
+
+      return false;
+    }
+
+    private boolean matchesScope(String actualScope, Set<String> scopes) {
+      if (scopes.isEmpty() || scopes.contains(actualScope)) {
+        return true;
+      }
+
+      for (String scope : scopes) {
+        if (actualScope.startsWith(scope)) {
+          return true;
+        }
+      }
+
+      return false;
+    }
+  }
+
+  private static class SparkMetricResult<T> implements MetricResult<T> {
```

Comment:
Use `@AutoValue` to simplify it (something like https://github.com/apache/beam/blob/a2c342cfd2c2f86593e6d4f30c25b8054fa2ffa2/runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectMetrics.java#L226)


I'd hesitate to put it in the SDK in this form since it discards the distinction between committed and attempted.